pkg/controller/clusterset/node.go

```diff
@@ -0,0 +1,134 @@
+package clusterset
+
+import (
+	"context"
+
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	"github.com/rancher/k3k/pkg/controller/util"
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	nodeController = "k3k-node-controller"
+)
+
+type NodeReconciler struct {
+	Client      ctrlruntimeclient.Client
+	Scheme      *runtime.Scheme
+	ClusterCIDR string
+}
+
+// Add adds a new controller to the manager
+func AddNodeController(ctx context.Context, mgr manager.Manager, clusterCIDR string) error {
+	// initialize a new Reconciler
+	reconciler := NodeReconciler{
+		Client:      mgr.GetClient(),
+		Scheme:      mgr.GetScheme(),
+		ClusterCIDR: clusterCIDR,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Node{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: maxConcurrentReconciles,
+		}).
+		Complete(&reconciler)
+}
+
+func (n *NodeReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	if n.ClusterCIDR != "" {
+		return reconcile.Result{}, nil
+	}
+
+	var clusterSetList v1alpha1.ClusterSetList
+	if err := n.Client.List(ctx, &clusterSetList); err != nil {
+		return reconcile.Result{}, util.LogAndReturnErr("failed to list clusterSets", err)
+	}
+
+	if len(clusterSetList.Items) <= 0 {
+		return reconcile.Result{}, nil
+	}
+
+	var node v1.Node
+	if err := n.Client.Get(ctx, types.NamespacedName{Name: req.Name}, &node); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return reconcile.Result{}, util.LogAndReturnErr("failed to handle get node", err)
+		}
+	}
+	if node.DeletionTimestamp.IsZero() {
+		for _, cs := range clusterSetList.Items {
+			klog.Infof("Add new CIDR %v of node %s to clusterSet %s networkpolicy", node.Spec.PodCIDR, req.Name, cs.Name)
+			if err := n.addCIDR(ctx, cs, node.Spec.PodCIDR); err != nil {
+				return reconcile.Result{}, util.LogAndReturnErr("failed to update networkpolicy with new node cidr", err)
+			}
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	for _, cs := range clusterSetList.Items {
+		klog.Infof("Remove CIDR %v of node %s from clusterSet %s networkpolicy", node.Spec.PodCIDR, req.Name, cs.Name)
+		if err := n.removeCIDR(ctx, cs, node.Spec.PodCIDR); err != nil {
+			return reconcile.Result{}, util.LogAndReturnErr("failed to remove cidr from networkpolicy", err)
```

Comment:
We seem to be mixing error logging and error wrapped logging together here. Is that intended?

Reply:
@MbolotSuse correct me if I am wrong but thats because Reconcile() in controller-runtime doesn't really log the returned error, however it begin a new cycle with a backoff for the controller, thats why I have a util.LogAndReturnErr() function to be able to log the error before returning it, that said I am open to what the controller-runtime does as a best practice for logging

Reply:
I think that controller runtime will log an error if reconcile fails: https://github.com/kubernetes-sigs/controller-runtime/blob/c50f2dfd0b618220108d320e3e690340a9246a75/pkg/internal/controller/controller.go#L329. For that reason I think we can just return the error here. 

That being said, the kubebuilder book contains a sample that logs and returns the error: https://book.kubebuilder.io/cronjob-tutorial/controller-implementation#2-list-all-active-jobs-and-update-the-status. Overall, I think that logging something specific with a returned error is fine. I think that @briandowns is right that doing both through a side function is odd, I'd prefer that we logged out and returned in two different statements, despite the fact that it produces more code.

Reply:
That makes sense, I think controller-runtime has a guide https://github.com/kubernetes-sigs/controller-runtime/blob/main/TMP-LOGGING.md, I will try to follow this guide and adjust the logging through the whole repo, but this might be useful in a separate PR

Reply:
I have added this issue https://github.com/rancher/k3k/issues/127 to track the logging problem in a separate pr(s)